File: app/src/main/java/ua/te/hackathon/smartcity2015/ui/main/MainActivity.java
package ua.te.hackathon.smartcity2015.ui.main;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import ua.te.hackathon.smartcity2015.R;
import ua.te.hackathon.smartcity2015.ui.BaseActivity;
import ua.te.hackathon.smartcity2015.ui.main.edit.EventCreationActivity;
import ua.te.hackathon.smartcity2015.ui.main.events.browse.BrowseEventsFragment;

public class MainActivity extends BaseActivity implements MainView {

  private static MainPresenter presenter;

  @Bind(R.id.toolbar)
  Toolbar toolbar;

  @Override
  public void onCreateAuthenticated(Bundle savedInstanceState) {
    setContentView(R.layout.activity_main);

    ButterKnife.bind(this);

    setSupportActionBar(toolbar);

    getSupportActionBar().setTitle(R.string.upcoming_games);

    getSupportFragmentManager().beginTransaction()
        .replace(R.id.container, BrowseEventsFragment.newInstance())
        .commit();
  }

  @OnClick(R.id.fab)
  public void onCreateNewEvent() {
    startActivity(EventCreationActivity.startActivity(getApplication()));
  }

  @Override
  protected void onPostCreate(@Nullable Bundle savedInstanceState) {
    super.onPostCreate(savedInstanceState);

    if (presenter == null) {
      presenter = new MainPresenter(getApplicationContext());
    }

    presenter.attachView(this);
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    ButterKnife.unbind(this);

    if (presenter != null) {
      presenter.detachView();

      if (isFinishing()) {
        presenter.onDestroy();
      }
    }
  }
}

File: app/src/main/java/ua/te/hackathon/smartcity2015/ui/main/MainActivity.kt
package ua.te.hackathon.smartcity2015.ui.main

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import ua.te.hackathon.smartcity2015.R
import ua.te.hackathon.smartcity2015.ui.BaseActivity
import ua.te.hackathon.smartcity2015.ui.intro.IntroActivity
import ua.te.hackathon.smartcity2015.ui.main.edit.EventCreationActivity
import ua.te.hackathon.smartcity2015.ui.main.events.browse.BrowseEventsFragment
import ua.te.hackathon.smartcity2015.utils.LoggedUser

class MainActivity : BaseActivity(), MainView {

  public override fun onCreateAuthenticated(savedInstanceState: Bundle?) {
    setContentView(R.layout.activity_main)

    setSupportActionBar(toolbar)

    fab.setOnClickListener({ v -> onCreateNewEvent() })

    supportActionBar!!.setTitle(R.string.upcoming_games)

    supportFragmentManager.beginTransaction().replace(R.id.container, BrowseEventsFragment.newInstance()).commit()
  }

  fun onCreateNewEvent() {
    startActivity(EventCreationActivity.startActivity(application))
  }

  override fun onPostCreate(savedInstanceState: Bundle?) {
    super.onPostCreate(savedInstanceState)

    if (presenter == null) {
      presenter = MainPresenter(applicationContext)
    }

    presenter!!.attachView(this)
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu_profile, menu)
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.actionLogout -> {
        LoggedUser.logOut()
        val intent = IntroActivity.startRegistration(this)
        startActivity(intent)
        finish()
      }
      else -> return super.onOptionsItemSelected(item)
    }
    return true
  }

  override fun onDestroy() {
    super.onDestroy()

    EventBus.getDefault().unregister(this)

    if (presenter != null) {
      presenter!!.detachView()

      if (isFinishing) {
        presenter!!.onDestroy()
        presenter = null
      }
    }
  }

  companion object {
    private var presenter: MainPresenter? = null
  }
}
